### PR TITLE
Connector example

### PR DESCRIPTION
**Pull request by zeerorg**
This PR updates the connector example with swarm and kubernetes deployment files.

How it is tested:
1. Clone this Pull request locally.
2. Go into the directory: `cd ./connector-sdk/cmd/tester/yaml`
3. For openfaas deployed on docker swarm do: `docker stack deploy func -c ./docker-compose.yml`
4. For openfaas deployed on kubernetes do: `kubectl create -f ./kubernetes --namespace openfaas`

Run this command to deploy a sample function and see `trigger-func` invocation count growing in ui.

```bash
faas-cli deploy --image=functions/nodeinfo --name=trigger-func --annotation topic=faas-request
```

Signed-off-by: Rishabh Gupta <r.g.gupta@outlook.com>

### CODE REVIEW

File: cmd/tester/main.go [hunk 1]
```diff
@@ -6,27 +6,56 @@ package main
...
 		data := []byte("test " + time.Now().String())
-		controller.Invoke("test", &data)
+
+		for _, topic := range topics {
+			controller.Invoke(topic, &data)
+		}
+	}
+}
+
+func getControllerConfig() *types.ControllerConfig {
```
Comment (alexellis) [inline cmd/tester/main.go]: If running in a the default is `gateway.openfaas:8080` / `gateway:8080` on Swarm. We should just not default at all if not passed and return an error `(types.ControllerConfig, error)`

File: cmd/tester/main.go [hunk 2]
```diff
@@ -6,27 +6,56 @@ package main
...
+	gURL, ok := os.LookupEnv("gateway_url")
+	if !ok {
+		gURL = "http://127.0.0.1:8080/"
 	}
+	fmt.Println("Gateway Url: ", gURL)
```
Comment (alexellis) [inline cmd/tester/main.go]: Let's not print this here.

File: cmd/tester/main.go [hunk 3]
```diff
@@ -6,27 +6,56 @@ package main
...
 
 	controller := types.NewController(creds, config)
 	fmt.Println(controller)
 	controller.BeginMapBuilder()
 
+	topics := getTopics()
+
 	// Simulate events emitting from queue/pub-sub
 	for {
 		time.Sleep(2 * time.Second)
```
Comment (alexellis) [inline cmd/tester/main.go]: Make this configurable to please.

File: cmd/tester/main.go [hunk 4]
```diff
@@ -6,27 +6,56 @@ package main
...
-	config := &types.ControllerConfig{
-		RebuildInterval: time.Millisecond * 1000,
-		GatewayURL:      "http://127.0.0.1:8080",
-		PrintResponse:   true,
-	}
+	creds := types.GetCredentials() // Get credentials for gateway login
+	config := getControllerConfig()
 
 	controller := types.NewController(creds, config)
 	fmt.Println(controller)
```
Comment (alexellis) [inline cmd/tester/main.go]: Remove this line.

Comment (alexellis) [inline cmd/tester/main.go]: Replace `fmt` with `log` for use in the container.

File: cmd/tester/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.9.2 as builder
```
Comment (alexellis) [inline cmd/tester/Dockerfile]: Copy the example from the vcenter connector please it has an updated Golang version.

File: cmd/tester/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+WORKDIR /go/src/github.com/openfaas-incubator/connector
+
+COPY main.go    .
+
+RUN go get -d -v ./...
+
+# Stripping via -ldflags "-s -w" 
+RUN CGO_ENABLED=0 GOOS=linux go build -a -ldflags "-s -w" -installsuffix cgo -o ./connector
+
+FROM alpine
```
Comment (alexellis) [inline cmd/tester/Dockerfile]: Copy the example from the vcenter connector please it has an updated alpine version, we must always pin to `3.8` or whatever version we are targeting.

File: cmd/tester/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.9.2 as builder
+RUN mkdir -p /go/src/github.com/openfaas-incubator/connector
+WORKDIR /go/src/github.com/openfaas-incubator/connector
```
Comment (alexellis) [inline cmd/tester/Dockerfile]: The path I guess is `connector-sdk` right? To match the repo.

File: cmd/tester/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+COPY main.go    .
+
+RUN go get -d -v ./...
```
Comment (alexellis) [inline cmd/tester/Dockerfile]: Please remove `go get`

It should use vendoring via `dep` or the correct path.

File: cmd/tester/README.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# Openfaas Sample Connector
+
+This folder contains a sample openfaas connector. You can use this as a base for creating your own connectors.
+For a more complex example checkout [kafka-connector](https://github.com/openfaas-incubator/kafka-connector)
+
+## How to Use
+
+1. Clone this repository: `git clone https://github.com/openfaas-incubator/connector-sdk.git`
+2. Go into the directory: `cd ./connector-sdk/cmd/tester/yaml`
+3. For openfaas deployed on docker swarm do: `docker stack deploy func -c ./docker-compose.yml`
```
Comment (alexellis) [inline cmd/tester/README.md]: `openfaas` -> OpenFaaS

Comment (alexellis) [inline cmd/tester/README.md]: `docker swarm` -> Docker Swarm

File: cmd/tester/Dockerfile (hunk 7 above)
Comment (alexellis) [reply cmd/tester/Dockerfile]: Actually I think it's:

`/go/src/github.com/openfaas-incubator/connector-sdk/cmd/tester/` correct?

**Comment by zeerorg** [comment]
I've worked through all the suggestions. This should be good to go.

Some major changes:
1. I removed fetching topics from environment variable (made the example hard to follow). now a single topic is defined.
2. Also removed print_response environment variable was not doing anything

File: cmd/tester/main.go [hunk 10]
```diff
@@ -5,28 +5,44 @@
...
+		time.Sleep(invokeTime)
 		data := []byte("test " + time.Now().String())
-		controller.Invoke("test", &data)
+
+		controller.Invoke(topic, &data)
+	}
+}
+
+func getControllerConfig() (*types.ControllerConfig, error) {
+	gURL, ok := os.LookupEnv("gateway_url")
```
Comment (alexellis) [inline cmd/tester/main.go]: Change to `gatewayURL`

File: cmd/tester/yaml/kubernetes/connector-dep.yml [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  template:
+    metadata:
+      labels:
+        app: connector
+        component: sample-connector
+    spec:
+      containers:
+      - name: sample-connector
+        image: zeerorg/connector-sample:1.0
+        env:
```
Comment (alexellis) [inline cmd/tester/yaml/kubernetes/connector-dep.yml]: Please add topics as an env-var, this is how connectors are intended to work.

**Comment by alexellis** [comment]
The scope of this PR was simply to provide a Dockerfile and Kubernetes YAML, not to change the example behaviour or functionality. Please can you revert the below?

> Some major changes:
> I removed fetching topics from environment variable (made the example hard to follow). now a single topic is defined.
> Also removed print_response environment variable was not doing anything

I think the extracting the configuration loading is fine to keep in the changes, but in a separate PR.

Alex

File: cmd/tester/main.go (hunk 3 above)
Comment (zeerorg) [reply cmd/tester/main.go]: Should I add an env-var to make this configurable ? Or a simple variable is enough ?
